design/healcheck-endpoint.md

```diff
@@ -0,0 +1,169 @@
+
+After many discussions, we decided to remove from the initial design the following items:
+- External IdP and User Federation checks. Including any of those checks makes the end-user situation worse.
+If we include an External IdP check in the Health endpoint, Kubernetes will plug all the pods out of the Service
+(a load balancer) if that IdP is down. Once all pods are removed from the Service, an end-user will see an HTTP 503
+Error page generated by the OpenShift router. If however, the External IdP is down, and we don't include it in the
+Health endpoint, a user will probably face an error during log in. But the error page will be displayed by Keycloak,
+and we control what message we send there. I consider this as a slightly better situation, although the end-result is
+the same - a user cannot log in.
+- Parametrized checks, e.g. `http://localhost:8080/auth/health?timeout=2s&fail-on=db,idp-google`. After removing
+the External IdPs and User Federation, the parametrized check is no longer necessary. Perhaps, we'll get back
+to it in the future.
+
+## Healthcheck parameters
+
+The new Health Endpoint need to support the `timeout` query parameter in the endpoint in the following form:
+
+    http://localhost:8080/auth/health?timeout=2s
+
+The optional `timeout` parameter causes the request to get terminated if not performed within a timeout.
+The implementation needs to take special care of releasing the thread pool (if necessary).
+Subsequent calls that take longer than expected should not exhaust the thread pools inside Wildfly. If not specified,
+`10s` will be used as default (the same timeout as the one used in Kubernetes).
+
+## Implementation
+
+From the implementation perspective, there are two ways of implementing Healthcheck calls within Keycloak:
+* Active - whenever we check if a database is down (this is just an example), we initiate a new connection and
+check if it succeeds.
+* Passive - we position the implementation based on Metrics components (see the Observability Design). We actively
+analyze database connection metrics reported by JCA subsystem
+(see [the currently available metrics in Keycloak](https://docs.google.com/spreadsheets/d/1YGn4wwUupgQsFTmlm-5VXgLYkigTdy_TEnDhqFxNXXo/edit#gid=0))
+and if we discover  increasing failure count, we report an at Health Endpoint as well.
+
+The former approach seems more natural but has a lof of drawbacks, including:
+* Additional connections allocated in the thread pools.
+* Often requires integrating with additional subsystems (e.g. JCA for database checks).
+* Complicated to ensure that `timeout` parameter is properly respected.
+
+All the above pain points are not applicable for the Passive Mode. Essentially, all database, User Federation LDAP servers
+and External IdPs probes are free of charge as we report success/failure calls for them in the Metrics components.
+With this approach, the Healthcheck Endpoint becomes an alternative reporting tool for the Metrics.
```

Comment:
This section doesn't seem to have any relevance

Reply:
I created this for @kenfinnigan. I wanted to make it crystal clear the implementation will use metrics and reports the status based on them. From the conceptual angle, it becomes a reporting tool. 

Let me leave it as is.